examples/tfidf.cpp

```diff
@@ -0,0 +1,156 @@
+    inputformat.set_auth(user, pwd);
+    load(inputformat, parse);
+    auto& term_list = husky::ObjListFactory::create_objlist<Term>();
+    husky::lib::Aggregator<int> totalDoc;
+    auto& ac = husky::lib::AggregatorFactory::get_channel();
+    auto& ch = husky::ChannelFactory::create_push_combined_channel<int, husky::SumCombiner<int>>(document_list, term_list);
+    auto& ch2 = husky::ChannelFactory::create_push_channel<std::pair<std::string, int>>(document_list, term_list);
+    list_execute(document_list, {}, {&ch, &ch2, &ac}, [&](Document& doc) {
+        doc.words.resize(0);
+        std::vector<int> count;
+        if (doc.content.size()> 0) {
+            boost::char_separator<char> sep(" \t\n.,()\'\":;!?<>");
+            boost::tokenizer<boost::char_separator<char>> tok(doc.content, sep);
+            for (auto &w : tok) {
+                doc.words.push_back(w);
+                std::transform(doc.words.back().begin(), doc.words.back().end(), doc.words.back().begin(), ::tolower);
+            }
+            doc.total_words = doc.words.size();
+            std::sort(doc.words.begin(), doc.words.end());
+            int n = 0;
+            for (int i = 0, j = 0; i < doc.words.size(); i = j) {
+                for (j = i + 1; j < doc.words.size(); j++) {
+                    if (doc.words.at(i).compare(doc.words.at(j)) != 0)
+                        break;
+                }
+                count.push_back(j - i);
+                doc.words[n++] = doc.words[i];
+            }
+            doc.words.resize(n);
+            doc.words.shrink_to_fit();
+            doc.tf.resize(doc.words.size());
+            doc.tf_idf.resize(doc.words.size());
+            for (int i = 0; i < doc.words.size(); i++) {
+                doc.tf.at(i) = static_cast<double>(count.at(i))/doc.total_words;
+            } 
+        }
+    });
+
+    husky::globalize(document_list);
+    list_execute(document_list, {}, {&ch, &ch2, &ac}, [&](Document& doc) {
+        for (int i = 0; i < doc.words.size(); i++) {
+            ch.push(1, doc.words.at(i));
+            ch2.push(std::make_pair(doc.id(), i), doc.words.at(i));
+        }
+        totalDoc.update(1);
+    });
+    auto& ch3 = husky::ChannelFactory::create_push_channel<std::pair<int, float>>(term_list, document_list);
+    list_execute(term_list,{&ch}, {}, [&](Term& t){
+        t.idf = log(static_cast<double>(totalDoc.get_value())/ch.get(t));
+    });
+    list_execute(term_list,{&ch2}, {&ch3},  [&](Term& t){
+        auto& msgs = ch2.get(t);
+        for (int i = 0; i < msgs.size(); i++){
+            std::pair<int, float> push_msg = std::make_pair(int(msgs[i].second), t.idf);
+            ch3.push(push_msg, msgs[i].first);
+        }
+    });
+    list_execute(document_list,{&ch3}, {}, [&](Document& doc) {
+        auto& msgs = ch3.get(doc);
+        for (int j = 0; j < msgs.size(); j++) {
+            int i = int(msgs[j].first);
+            doc.tf_idf.at(i) = doc.tf.at(i)*msgs[j].second;
```

Comment:
In the first list_execute (line 100 - 108), I have deleted all the duplicated words and only non-duplicated words were stored in words vector for one document. So one document only send no more than one message to one term through ch2 and thus one term send no more than one message to one document through ch3. Therefore, I do not think that same words in the same documents will cause more calculation.

I will fix the other problems asap.